File: library/src/main/java/com/hacknife/immersive/Immersive.java
package com.hacknife.immersive;

import android.app.Activity;
import android.os.Build;
import android.os.Build.VERSION;
import android.support.annotation.LayoutRes;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;


import static android.view.View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION;
import static android.view.Window.FEATURE_NO_TITLE;
import static android.view.WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION;
import static android.view.WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS;


/**
 * author  : Hacknife
 * e-mail  : 4884280@qq.com
 * github  : http://github.com/hacknife
 * project : Immersive
 */
public class Immersive {
    private static final String TAG = Immersive.class.getName();

    public Immersive() {
    }

    private static void compatible(Activity activity) {
        if (VERSION.SDK_INT >= 19 && VERSION.SDK_INT < 21) {
            activity.getWindow().requestFeature(FEATURE_NO_TITLE);
            activity.getWindow().addFlags(FLAG_TRANSLUCENT_STATUS);
            activity.getWindow().addFlags(FLAG_TRANSLUCENT_NAVIGATION);
        }
        if (VERSION.SDK_INT >= 21) {
            activity.getWindow().addFlags(SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION);
        }
    }

    public static void setContentView(Activity activity, @LayoutRes int layoutRes, int statusRes, int navigationRes, boolean statusEmbed, boolean navigationEmbed) {
        compatible(activity);
        Orientation orientation = ImmersiveHelper.getActivityOrientationForContext(activity);
        if (orientation== Orientation._0 || orientation == Orientation._180){
            activity.setContentView(R.layout.activity_immersive_vertical);
        }else  if (orientation == Orientation._90){
            activity.setContentView(R.layout.activity_immersive_horizontal_90);
        } else {
            activity.setContentView(R.layout.activity_immersive_horizontal_270);
        }
        StatusView statusView = activity.findViewById(R.id.immersive_status);
        FrameLayout content = activity.findViewById(R.id.immersive_content);
        NavigationView navigationView = activity.findViewById(R.id.immersive_navigation);
        statusView.setBackgroundResource(statusRes);
        navigationView.setBackgroundResource(navigationRes);
        if (statusEmbed) statusView.setVisibility(View.GONE);
        if (navigationEmbed) navigationView.setVisibility(View.GONE);
        content.addView(LayoutInflater.from(activity).inflate(layoutRes, null));
    }

    public static void hideStatus(Activity activity) {
        StatusView statusView = activity.findViewById(R.id.immersive_status);
        if (statusView != null) statusView.setVisibility(View.GONE);
    }

    public static boolean isShowOfStatus(Activity activity) {
        StatusView statusView = activity.findViewById(R.id.immersive_status);
        if (statusView != null) return statusView.getVisibility() == View.VISIBLE;
        return false;
    }

    public static boolean isShowOfNavigation(Activity activity) {
        NavigationView navigationView = activity.findViewById(R.id.immersive_navigation);
        if (navigationView != null) return navigationView.getVisibility() == View.VISIBLE;
        return false;
    }


    public static void showStatus(Activity activity) {
        StatusView statusView = activity.findViewById(R.id.immersive_status);
        if (statusView != null) statusView.setVisibility(View.VISIBLE);
    }

    public static void hideNavigation(Activity activity) {
        NavigationView navigationView = activity.findViewById(R.id.immersive_navigation);
        if (navigationView != null) navigationView.setVisibility(View.GONE);
    }

    public static void showNavigation(Activity activity) {
        NavigationView navigationView = activity.findViewById(R.id.immersive_navigation);
        if (navigationView != null) navigationView.setVisibility(View.VISIBLE);
    }


    public static void setContentBackgroundColor(Activity activity, int color) {
        View content = activity.findViewById(R.id.immersive_content);
        if (content != null) {
            content.setBackgroundColor(color);
        }
    }

    public static void setContentBackgroundColorResource(Activity activity, int colorRes) {
        View content = activity.findViewById(R.id.immersive_content);
        if (content != null) {
            content.setBackgroundResource(colorRes);
        }
    }


    public static void setStatusBarColor(Activity activity, int color) {
        View status = activity.findViewById(R.id.immersive_status);
        if (status != null) {
            status.setBackgroundColor(color);
        }

    }

    public static void setStatusBarColorResource(Activity activity, int colorRes) {
        View status = activity.findViewById(R.id.immersive_status);
        if (status != null) {
            status.setBackgroundResource(colorRes);
        }
    }

    public static void setNavigationBarColor(Activity activity, int color) {
        View navigation = activity.findViewById(R.id.immersive_navigation);
        if (navigation != null) {
            navigation.setBackgroundColor(color);
        }
    }

    public static void setNavigationBarColorResource(Activity activity, int colorRes) {
        View navigation = activity.findViewById(R.id.immersive_navigation);
        if (navigation != null) {
            navigation.setBackgroundResource(colorRes);
        }
    }

    public static boolean setStatusContentColor(Activity activity, MODE mode) {
        if (VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (mode == MODE.BLACK)
                activity.getWindow()
                        .getDecorView()
                        .setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);

            else if (mode == MODE.WHITE)
                activity.getWindow()
                        .getDecorView()
                        .setSystemUiVisibility(View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN | View.SYSTEM_UI_FLAG_LAYOUT_STABLE);
            return true;
        } else {
            return false;
        }
    }

    public static boolean setNavigationContentColor(Activity activity, MODE mode) {
        if (VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (mode == MODE.BLACK)
                activity.getWindow().clearFlags(FLAG_TRANSLUCENT_NAVIGATION);
            else if (mode == MODE.WHITE)
                activity.getWindow().addFlags(FLAG_TRANSLUCENT_NAVIGATION);
            return true;
        } else {
            return false;
        }
    }

    public static View layoutView(Activity activity) {
        ViewGroup group = activity.<ViewGroup>findViewById(R.id.immersive_content);
        if (group != null) return group.getChildAt(0);
        else return null;
    }

    public enum MODE {
        WHITE, BLACK
    }
}

File: library/src/main/java/com/hacknife/immersive/Immersive.kt
package com.hacknife.immersive

import android.app.Activity
import android.graphics.Color
import android.os.Build
import android.os.Build.VERSION
import android.view.*
import android.widget.FrameLayout
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.FragmentActivity

/**
 * author  : Hacknife
 * e-mail  : 4884280@qq.com
 * github  : http://github.com/hacknife
 * project : Immersive
 */


private val stateMap = hashMapOf<String, State>()

private fun compatible19(activity: FragmentActivity) {
    if (VERSION.SDK_INT >= 19 && VERSION.SDK_INT < 21) {
        activity.window.requestFeature(Window.FEATURE_NO_TITLE)
        activity.window
            .addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
        activity.window
            .addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
    }

}

private fun compatible21(activity: FragmentActivity) {
    if (VERSION.SDK_INT >= 21) {
        activity.window.statusBarColor = Color.TRANSPARENT
        activity.window.navigationBarColor = Color.TRANSPARENT
        activity.window?.decorView?.systemUiVisibility =
            systemUiFlag(true, true)
        stateMap.put(activity.hashCode().toString(), State(true, true))
    }
}

private fun systemUiFlag(statusBarDark: Boolean, navigationBarDark: Boolean): Int {
    var flag = (View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
            or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
            or View.SYSTEM_UI_FLAG_LAYOUT_STABLE)
    when {
        VERSION.SDK_INT >= Build.VERSION_CODES.O -> {
            if (statusBarDark) flag = flag or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
            if (navigationBarDark) flag = flag or View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
        }
        VERSION.SDK_INT >= Build.VERSION_CODES.M -> {
            if (statusBarDark) flag = flag or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
        }
    }
    return flag
}


fun AppCompatActivity.setContentView(
    @LayoutRes layoutRes: Int,
    @ColorRes statusRes: Int,
    @ColorRes navigationRes: Int,
    statusEmbed: Boolean,
    navigationEmbed: Boolean
) {
    compatible19(this)
    val orientation =
        ImmersiveHelper.getActivityOrientationForContext(this)
    if (orientation == Orientation._0 || orientation == Orientation._180) {
        this.setContentView(R.layout.activity_immersive_vertical)
    } else if (orientation == Orientation._90) {
        this.setContentView(R.layout.activity_immersive_horizontal_90)
    } else {
        this.setContentView(R.layout.activity_immersive_horizontal_270)
    }
    compatible21(this)
    val statusView: StatusView = this.findViewById(R.id.immersive_status)
    val content = this.findViewById<FrameLayout>(R.id.immersive_content)
    val navigationView: NavigationView = this.findViewById(R.id.immersive_navigation)
    statusView.setBackgroundResource(statusRes)
    navigationView.setBackgroundResource(navigationRes)
    if (statusEmbed) statusView.visibility = View.GONE
    if (navigationEmbed) navigationView.visibility = View.GONE
    content.addView(LayoutInflater.from(this).inflate(layoutRes, null))
}


fun Activity.hideStatus() {
    val statusView: StatusView? = findViewById(R.id.immersive_status)
    if (statusView != null) statusView.visibility = View.GONE
}


fun Activity.isShowOfStatus(): Boolean {
    val statusView: StatusView? = findViewById(R.id.immersive_status)
    return if (statusView != null) statusView.visibility == View.VISIBLE else false
}


fun Activity.isShowOfNavigation(): Boolean {
    val navigationView: NavigationView? = findViewById(R.id.immersive_navigation)
    return if (navigationView != null) navigationView.visibility == View.VISIBLE else false
}


fun Activity.showStatus() {
    val statusView: StatusView? = findViewById(R.id.immersive_status)
    if (statusView != null) statusView.visibility = View.VISIBLE
}


fun Activity.hideNavigation() {
    val navigationView: NavigationView? = findViewById(R.id.immersive_navigation)
    if (navigationView != null) navigationView.visibility = View.GONE
}


fun Activity.showNavigation() {
    val navigationView: NavigationView? = findViewById(R.id.immersive_navigation)
    if (navigationView != null) navigationView.visibility = View.VISIBLE
}

fun Activity.setContentBackgroundColor(@ColorInt color: Int) {
    val content = findViewById<View>(R.id.immersive_content)
    content?.setBackgroundColor(color)
}

fun Activity.setContentBackgroundColorResource(@ColorRes colorRes: Int) {
    val content = findViewById<View>(R.id.immersive_content)
    content?.setBackgroundResource(colorRes)
}


fun Activity.setStatusBarColor(@ColorInt color: Int) {
    val status = findViewById<View>(R.id.immersive_status)
    status?.setBackgroundColor(color)
}

fun Activity.setStatusBarColorResource(@ColorRes colorRes: Int) {
    val status = findViewById<View>(R.id.immersive_status)
    status?.setBackgroundResource(colorRes)
}


fun Activity.setNavigationBarColor(@ColorInt color: Int) {
    val navigation = findViewById<View>(R.id.immersive_navigation)
    navigation?.setBackgroundColor(color)
}

fun Activity.setNavigationBarColorResource(@ColorRes colorRes: Int) {
    val navigation = findViewById<View>(R.id.immersive_navigation)
    navigation?.setBackgroundResource(colorRes)
}


fun Activity.setStatusContentColor(mode: MODE): Boolean {
    return if (VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        val navigationBar = stateMap[this.hashCode().toString()]?.navigationBar ?: true
        if (mode == MODE.WHITE) {
            this.window.decorView.systemUiVisibility = systemUiFlag(false, navigationBar)
            stateMap[this.hashCode().toString()]?.stateBar = false
        } else {
            this.window.decorView.systemUiVisibility = systemUiFlag(true, navigationBar)
            stateMap[this.hashCode().toString()]?.stateBar = true
        }
        true
    } else {
        false
    }
}


fun Activity.setNavigationContentColor(mode: MODE): Boolean {
    return if (VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        val stateBar = stateMap[this.hashCode().toString()]?.stateBar ?: true
        if (mode == MODE.WHITE) {
            this.window.decorView.systemUiVisibility = systemUiFlag(stateBar, false)
            stateMap[this.hashCode().toString()]?.navigationBar = false
        } else {
            this.window.decorView.systemUiVisibility = systemUiFlag(stateBar, true)
            stateMap[this.hashCode().toString()]?.navigationBar = true
        }
        true
    } else {
        false
    }
}

fun Activity.rootView(): View? {
    val group = findViewById<ViewGroup>(R.id.immersive_content)
    return group?.getChildAt(0)
}

enum class MODE {
    WHITE, BLACK
}
